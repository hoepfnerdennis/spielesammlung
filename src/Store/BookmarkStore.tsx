import { useEffect } from 'react';
import makeStore from '../utils/makeStore';
import useLocalStorage from '../utils/useLocalStoreage';

interface IAction {
  type: string;
  payload: any;
}

export enum ActionTypes {
  'SET_BOOKMARKS' = 'SET_BOOKMARKS',
  'ADD_BOOKMARK' = 'ADD_BOOKMARK',
  'REMOVE_BOOKMARK' = 'REMOVE_BOOKMARK',
}

const initialState: string[] = localStorage.getItem('bookmarks')?.split(',') || [];

const dataReducer = (state: string[], action: IAction): string[] => {
  switch (action.type) {
    case ActionTypes.SET_BOOKMARKS:
      return action.payload;
    case ActionTypes.ADD_BOOKMARK:
      return [...state, action.payload];
    case ActionTypes.REMOVE_BOOKMARK:
      return state.filter((value) => value !== action.payload);
    default:
      return state;
  }
};

const [BookmarksProvider, useBookmarks, useDispatch] = makeStore<string[], IAction>(
  dataReducer,
  initialState
);

const useMark = (): ((id: string) => void) => {
  const dispatch = useDispatch();
  const bookmarks = useBookmarks();
  const { setItem } = useLocalStorage();

  const mark = (id: string): void => {
    if (bookmarks.includes(id)) {
      return dispatch({ type: ActionTypes.REMOVE_BOOKMARK, payload: id });
    }
    return dispatch({ type: ActionTypes.ADD_BOOKMARK, payload: id });
  };

  useEffect(() => {
    setItem<string[]>('bookmarks', bookmarks);
  }, [bookmarks, setItem]);

  return mark;
};

export { BookmarksProvider, useBookmarks, useMark };
