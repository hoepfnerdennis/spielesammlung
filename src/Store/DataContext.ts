import { createContext } from 'react';
import { IGame, SetFilterFunc, ActiveFiltersMap } from './types';

const defaultValue = {
  games: [],
  setFilter: () => (): void => {
    // default
  },
  activeFilters: new Map(),
};

export default createContext<{
  games: IGame[];
  setFilter: SetFilterFunc;
  activeFilters: ActiveFiltersMap;
}>(defaultValue);
