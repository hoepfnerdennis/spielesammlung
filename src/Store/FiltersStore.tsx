import makeStore from '../utils/makeStore';
import { ActiveFiltersMap } from './types';

interface IAction {
  type: string;
  payload: ActiveFiltersMap;
}

export enum ActionTypes {
  'SET_FILTER' = 'SET_FILTER',
}

const initialState: ActiveFiltersMap = new Map();

const dataReducer = (state: ActiveFiltersMap, action: IAction): ActiveFiltersMap => {
  switch (action.type) {
    case ActionTypes.SET_FILTER:
      return action.payload;
    default:
      return state;
  }
};

const [FiltersProvider, useFilters, useDispatch] = makeStore<ActiveFiltersMap, IAction>(
  dataReducer,
  initialState
);

export { FiltersProvider, useFilters, useDispatch };
