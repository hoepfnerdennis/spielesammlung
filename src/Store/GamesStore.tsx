import makeStore from '../utils/makeStore';
import { IGame } from './types';

interface IAction {
  type: string;
  payload: IGame[];
}

export enum ActionTypes {
  'SET_GAMES' = 'SET_GAMES',
}

const initialState: IGame[] = [];

const dataReducer = (state: IGame[], action: IAction): IGame[] => {
  switch (action.type) {
    case ActionTypes.SET_GAMES:
      return action.payload;
    default:
      return state;
  }
};

const [GamesProvider, useGames, useDispatch] = makeStore<IGame[], IAction>(
  dataReducer,
  initialState
);

export { GamesProvider, useGames, useDispatch };
