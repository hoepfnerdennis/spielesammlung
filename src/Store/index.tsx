import React, { createContext, useReducer } from 'react';
import { IState, Reducer, IAction } from './types';

export const FETCH_DATA_FINISHED = 'FETCH_DATA_FINISHED';
export const SET_SEARCH_TERM = 'SET_SEARCH_TERM';
export const SET_PALYER_FROM = 'SET_PALYER_FROM';
export const SET_PLAYER_TO = 'SET_PLAYER_TO';
export const SET_FAVORITE = 'SET_FAVORITE';

const initialState: IState = {
  games: [],
  filters: {
    playersFrom: undefined,
    playersTo: undefined,
    favorite: undefined,
    search: undefined,
  },
};

export const Store = createContext<IState | any>(initialState);

const reducer = (state: IState, action: IAction): IState => {
  switch (action.type) {
    case FETCH_DATA_FINISHED:
      return { ...state, games: action.payload };

    case SET_PALYER_FROM:
      return { ...state, filters: { ...state.filters, playersFrom: action.payload } };

    case SET_PLAYER_TO:
      return { ...state, filters: { ...state.filters, playersTo: action.payload } };

    case SET_SEARCH_TERM:
      return { ...state, filters: { ...state.filters, search: action.payload } };

    case SET_FAVORITE:
      return { ...state, filters: { ...state.filters, favorite: action.payload } };

    default:
      return state;
  }
};

export const StoreProvider: React.SFC = ({ children }): JSX.Element => {
  const [state, dispatch] = useReducer<Reducer>(reducer, initialState);
  return <Store.Provider value={{ state, dispatch }}>{children}</Store.Provider>;
};
