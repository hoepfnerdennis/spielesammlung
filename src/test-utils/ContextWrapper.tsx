import React, { FunctionComponent } from 'react';
import { createRenderer } from 'fela';
import { RendererProvider } from 'react-fela';

const renderer = createRenderer();

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const ContextWrapper: FunctionComponent = ({ children }) => {
  return <RendererProvider renderer={renderer}>{children}</RendererProvider>;
};

export default ContextWrapper;
