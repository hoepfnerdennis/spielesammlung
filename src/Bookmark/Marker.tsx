import React from 'react';
import { useFela } from 'react-fela';
import { IStyle } from 'fela';
import { useBookmarks, useMark } from '../Store/BookmarkStore';

const invisibleButton: IStyle = {
  fontSize: '1.6rem',
  appearance: 'none',
  border: 'none',
  background: 'transparent',
};

const Marker: React.FC<{ id: string }> = ({ id }) => {
  const bookmarks = useBookmarks();
  const mark = useMark();
  const { css } = useFela();
  return (
    <button
      data-testid="bookmark-toggle"
      type="button"
      onClick={(): void => mark(id)}
      className={css(invisibleButton)}>
      {bookmarks.includes(id) ? '‚ù§Ô∏è' : 'ü§ç'}
    </button>
  );
};

export default Marker;
