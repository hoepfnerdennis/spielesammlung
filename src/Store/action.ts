import { useCallback, useEffect } from 'react';
import { IGame, IAPIResponse, FiltersConfig, FilterKey, FilterValue, SetFilterFunc } from './types';
import { mapResultsToGames } from '../utils';
import { useDispatch as useDispatchGames, ActionTypes as GamesActionTypes } from './GamesStore';
import {
  useFilters,
  useDispatch as useDispatchFilters,
  ActionTypes as FiltersActionTypes,
} from './FiltersStore';

const SPACE_ID = '9sxha2f3gm24';
const API_TOKEN = '7LDIC95TsrYOfZwEQnbAuMHtij97kfk5r1dIRiGqT8M';
// For local testing replace BASE_URI with
// const BASE_URI = 'http://localhost:3001/response.json';
const BASE_URI = `https://cdn.contentful.com/spaces/${SPACE_ID}/environments/master/entries?access_token=${API_TOKEN}`;

const filtersConfig: FiltersConfig = {
  playersFrom: '[lte]',
  playersTo: '[gte]',
  name: '[match]',
  favorite: '',
  simpleRules: '',
  drinkingGame: '',
};

const fetchGames = async (url: string) => {
  try {
    const response = await fetch(url);
    const data: IAPIResponse = await response.json();
    const gamesFromAPI: IGame[] = mapResultsToGames(data);
    return gamesFromAPI;
  } catch {
    return [];
  }
};

const fetchGamesAction = async (filters: Map<FilterKey, FilterValue>): Promise<IGame[]> => {
  let url = `${BASE_URI}&content_type=game`;
  filters.forEach((value, key) => {
    url += `&fields.${FilterKey[key]}${filtersConfig[FilterKey[key]]}=${value}`; // &fields.playersTo[gte]=${from}`;
  });
  return fetchGames(url);
};

export const useSetFilter = (): SetFilterFunc => {
  const dispatch = useDispatchFilters();
  const activeFilters = useFilters();

  const addFilter = useCallback(
    (key: FilterKey, value: FilterValue): void => {
      if (activeFilters.get(key) === value) {
        return;
      }
      const newMap = new Map(activeFilters);
      newMap.set(key, value);
      dispatch({ type: FiltersActionTypes.SET_FILTER, payload: newMap });
    },
    [activeFilters, dispatch]
  );

  const removeFilter = useCallback(
    (key: FilterKey): void => {
      if (!activeFilters.has(key)) {
        return;
      }
      const newMap = new Map(activeFilters);
      newMap.delete(key);
      dispatch({ type: FiltersActionTypes.SET_FILTER, payload: newMap });
    },
    [activeFilters, dispatch]
  );

  const setFilter = useCallback(
    (key: FilterKey) => (value?: FilterValue): void => {
      if (value) {
        addFilter(key, value);
      } else {
        removeFilter(key);
      }
    },
    [addFilter, removeFilter]
  );

  return setFilter;
};

export const useFetchGames = (): void => {
  const dispatch = useDispatchGames();
  const activeFilters = useFilters();
  useEffect(() => {
    const getGames = async (): Promise<void> => {
      const newGames = await fetchGamesAction(activeFilters);
      dispatch({ type: GamesActionTypes.SET_GAMES, payload: newGames });
    };
    getGames();
  }, [activeFilters, dispatch]);
};
