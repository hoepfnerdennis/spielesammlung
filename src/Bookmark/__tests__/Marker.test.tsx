import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Marker from '../Marker';
import { useBookmarks, useMark } from '../../Store/BookmarkStore';
import ContextWrapper from '../../test-utils/ContextWrapper';

jest.mock('../../Store/BookmarkStore', () => ({
  useBookmarks: jest.fn(),
  useMark: jest.fn(),
}));

describe('Marker', () => {
  it.each([
    ['active', ['a', 'b', 'c'], 'a'],
    ['inactive', ['a', 'b', 'c'], 'd'],
  ])('should render %s', (desc, bookmarks, id) => {
    const mark = jest.fn();
    // @ts-ignore
    useBookmarks.mockImplementation(() => bookmarks);
    // @ts-ignore
    useMark.mockImplementation(() => mark);

    const { container, getByTestId } = render(
      <ContextWrapper>
        <Marker id={id} />
      </ContextWrapper>
    );

    fireEvent.click(getByTestId('bookmark-toggle'));
    expect(mark).toHaveBeenCalledWith(id);
    expect(mark).toHaveBeenCalledTimes(1);

    expect(container).toMatchSnapshot();
  });
});
