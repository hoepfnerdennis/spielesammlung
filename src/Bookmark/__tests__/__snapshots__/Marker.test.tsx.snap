// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Marker should render active 1`] = `
<div>
  <button
    class="a b c d"
    data-testid="bookmark-toggle"
    type="button"
  >
    ❤️
  </button>
</div>
`;

exports[`Marker should render inactive 1`] = `
<div>
  <button
    class="a b c d"
    data-testid="bookmark-toggle"
    type="button"
  >
    🤍
  </button>
</div>
`;
